File: packages/dart/lib/src/criterion/read.dart
import '../error.dart';
import '../executor.dart';
import '../fields.dart';
import '../paths.dart';
import 'model.dart';

/// 从定义里的字段认出一条判据（不校验）。
Criterion criterionOf(Map map) {
  final description = textOf(map, 'description');
  final kind = textOf(map, 'executor');
  if (kind == agent) return AgentJudgement(description);
  if (kind == human) return HumanGate(description);
  final path = textOf(map, 'path');
  if (path.isNotEmpty) return PathExists(path, description: description);
  final absent = textOf(map, 'absent');
  if (absent.isNotEmpty) return PathAbsent(absent, description: description);
  final file = textOf(map, 'file');
  if (file.isNotEmpty) {
    return FileContains(file, textOf(map, 'contains'), description: description);
  }
  return CommandRun(textOf(map, 'run'), description: description);
}

/// 读一条判据：不是映射、取值不对、缺该有的字段，当场报错。
///
/// [step] 与 [order] 是这条判据的位置（第几个步骤、第几条判据），只用来说话；
/// 返回的是认好的值对象。
Criterion readCriterion(Object? value, int step, int order) {
  final at = CriterionPosition(step, order);
  // 先看是不是映射：不是映射时要说「不是映射」，不能先说 executor 该怎么写——
  // 那样报错会指错方向（2026-09-12 之前正是这个顺序，这条分支因此永远走不到）。
  if (value is! Map) {
    throw DefinitionError(at, const CriterionNotMapping());
  }
  final kind = textOf(value, 'executor');
  if (!criterionTypes.contains(kind)) {
    throw DefinitionError(at, const BadCriterionExecutor());
  }
  final odd = unknownFields(value, criterionFields);
  if (odd.isNotEmpty) {
    throw DefinitionError(at, UnknownCriterionFields(odd));
  }
  final given = [
    'path',
    'absent',
    'file',
    'contains',
    'run',
  ].where((name) => value[name] != null).toList();
  if (kind == rule) {
    if (given.isEmpty) {
      throw DefinitionError(at, const RuleNeedsJudgement());
    }
    if (given.contains('contains') && !given.contains('file')) {
      throw DefinitionError(at, const ContainsNeedsFile());
    }
    if (given.contains('file') && !given.contains('contains')) {
      throw DefinitionError(at, const FileNeedsContains());
    }
    final others = given
        .where((name) => name != 'file' && name != 'contains')
        .toList();
    if (others.length > 1 || (others.isNotEmpty && given.contains('file'))) {
      throw DefinitionError(at, const OnlyOneJudgement());
    }
  } else {
    if (textOf(value, 'description').isEmpty) {
      throw DefinitionError(at, NeedsDescription(kind));
    }
    if (given.isNotEmpty) {
      throw DefinitionError(at, NoRuleFields(kind, given));
    }
  }
  // 路径里只认四个占位；写别的（如 `{{foo}}`）算不合语法（规范 `process/workflow.md`·语法）。
  final unknown = <String>[];
  for (final field in ['path', 'absent', 'file']) {
    for (final name in placeholdersIn(textOf(value, field))) {
      if (!placeholderNames.contains(name) && !unknown.contains(name)) {
        unknown.add(name);
      }
    }
  }
  if (unknown.isNotEmpty) {
    throw DefinitionError(at, UnknownPlaceholder(unknown));
  }
  return criterionOf(value);
}

File: packages/rust/src/criterion/read.rs
//! 判据的读法：从定义里的字段读出 [`Criterion`]，顺带把语法过一遍。
//!
//! 判据是**字段**，不是一行小语法：`path` 存在、`absent` 不存在、
//! `file` + `contains` 含这段文字、`run` 这条命令退出码为零。字段名、取值、
//! 不认识、缺了、越界，当场报错。规矩的出处是 `docs/specification/process/workflow.md`·语法。

use super::model::Criterion;
use crate::executor::{AGENT, CRITERION_TYPES, HUMAN, RULE};
use crate::workflow::{DefinitionError, text_of, unknown_fields};
use serde_yaml::Value as Yaml;

/// 一条判据认得的字段。
const CRITERION_FIELDS: [&str; 7] = [
    "executor",
    "description",
    "path",
    "absent",
    "file",
    "contains",
    "run",
];

/// 从定义里的字段认出一条判据（不校验）。
pub fn criterion_of(value: &Yaml) -> Criterion {
    let description = text_of(value, "description");
    let kind = text_of(value, "executor");
    if kind == AGENT {
        return Criterion::AgentJudgement { description };
    }
    if kind == HUMAN {
        return Criterion::HumanGate { description };
    }
    let path = text_of(value, "path");
    if !path.is_empty() {
        return Criterion::PathExists { path, description };
    }
    let absent = text_of(value, "absent");
    if !absent.is_empty() {
        return Criterion::PathAbsent {
            absent,
            description,
        };
    }
    let file = text_of(value, "file");
    if !file.is_empty() {
        return Criterion::FileContains {
            file,
            contains: text_of(value, "contains"),
            description,
        };
    }
    Criterion::CommandRun {
        run: text_of(value, "run"),
        description,
    }
}

/// 读一条判据：取值不对、缺该有的字段，当场报错。
///
/// `file` 与 `place` 只用来说话；返回的是认好的值对象。
pub fn read_criterion(value: &Yaml, file: &str, place: &str) -> Result<Criterion, DefinitionError> {
    let kind = text_of(value, "executor");
    if !CRITERION_TYPES.contains(&kind.as_str()) {
        return Err(DefinitionError(format!(
            "{file} {place}的 executor 只能是 {}（谁判：规则引擎 / 智能体 / 人）",
            CRITERION_TYPES.join(" / ")
        )));
    }
    let criterion_map = value
        .as_mapping()
        .ok_or_else(|| DefinitionError(format!("{file} {place}不是映射")))?;
    let odd = unknown_fields(criterion_map, &CRITERION_FIELDS);
    if !odd.is_empty() {
        return Err(DefinitionError(format!(
            "{file} {place}有不认识的字段：{}（只认 {}）",
            odd.join("、"),
            CRITERION_FIELDS.join("、")
        )));
    }
    let given: Vec<&str> = ["path", "absent", "file", "contains", "run"]
        .into_iter()
        .filter(|name| value.get(*name).is_some())
        .collect();
    if kind == RULE {
        if given.is_empty() {
            return Err(DefinitionError(format!(
                "{file} {place}是 rule，得写一条判法（path / absent / file+contains / run）"
            )));
        }
        if given.contains(&"contains") && !given.contains(&"file") {
            return Err(DefinitionError(format!(
                "{file} {place}写了 contains，还得写 file"
            )));
        }
        if given.contains(&"file") && !given.contains(&"contains") {
            return Err(DefinitionError(format!(
                "{file} {place}写了 file，还得写 contains"
            )));
        }
        let others: Vec<&str> = given
            .iter()
            .copied()
            .filter(|name| *name != "file" && *name != "contains")
            .collect();
        if others.len() > 1 || (!others.is_empty() && given.contains(&"file")) {
            return Err(DefinitionError(format!(
                "{file} {place}的判法只能一种：path / absent / file+contains / run"
            )));
        }
    } else {
        if text_of(value, "description").is_empty() {
            return Err(DefinitionError(format!(
                "{file} {place}是 {kind}，必须写 description（判准 / 要人拍板的事）"
            )));
        }
        if !given.is_empty() {
            return Err(DefinitionError(format!(
                "{file} {place}是 {kind}，不该带 {}（那是 rule 的字段）",
                given.join("、")
            )));
        }
    }
    Ok(criterion_of(value))
}
